File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {

        int m = grid.length;
        int n = grid[0].length;
        int count = 0;

        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j] == '1'){
                    changeToLand(grid, i, j, m, n);
                    count++;
                }
            }
        }       

        return count; 
    }

    public void changeToLand(char[][] grid, int i, int j, int m, int n){
        if(i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == '0') return;

        grid[i][j] = '0';

        changeToLand(grid, i+1, j, m, n);
        changeToLand(grid, i-1, j, m, n);        
        changeToLand(grid, i, j+1, m, n);
        changeToLand(grid, i, j-1, m, n);                
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        count = 0
        m, n = len(grid)-1, len(grid[0])-1

        def dfs(i, j):
            if i < 0 or j < 0 or i > m or j > n or grid[i][j] == "0":
                return 
                
            grid[i][j] = "0"

            dfs(i-1, j)
            dfs(i+1, j)
            dfs(i, j-1)
            dfs(i, j+1)

        for i in range(m+1):
            for j in range(n+1):
                if grid[i][j] == "1":
                    dfs(i, j)
                    count += 1
            

        return count